tag/validator/builder.go

```diff
@@ -0,0 +1,169 @@
+package validator
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/go-playground/locales/en"
+	ut "github.com/go-playground/universal-translator"
+	"github.com/go-playground/validator/v10"
+	en_translation "github.com/go-playground/validator/v10/translations/en"
+)
+
+const defaultLocale = "en"
+
+// Builder is type to build validator
+type Builder struct {
+	fieldValidations  []FieldValidation
+	structValidations []StructValidation
+	translations      []Translation
+
+	validate   *validator.Validate
+	translator ut.Translator
+}
+
+// WithTranslations tells builder to include custom translation
+func (b *Builder) WithTranslations(translations []Translation) *Builder {
+	output := *b
+	output.translations = translations
+	return &output
+}
+
+// WithStructValidations tells builder to include custom struct validation
+func (b *Builder) WithStructValidations(structValidations []StructValidation) *Builder {
+	output := *b
+	output.structValidations = structValidations
+	return &output
+}
+
+// WithFieldValidations tells builder to include custom field validation
+func (b *Builder) WithFieldValidations(fieldValidations []FieldValidation) *Builder {
+	output := *b
+	output.fieldValidations = fieldValidations
+	return &output
+}
+
+// Build builds the validator
+func (b *Builder) Build() (Validator, error) {
+	b.translator = b.initializeTranslator()
+	validate, err := b.initializeValidate(b.translator)
+	if err != nil {
+		return nil, err
+	}
+	b.validate = validate
+
+	if b.fieldValidations != nil {
+		err = b.registerFieldValidations(b.validate, b.fieldValidations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if b.structValidations != nil {
+		b.registerStructValidations(b.validate, b.structValidations)
+	}
+	if b.translations != nil {
+		err = b.registerTranslations(b.validate, b.translator, b.translations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return b, nil
+}
+
+// Validate validates the type for any violation
+func (b *Builder) Validate(s interface{}) error {
+	err := b.validate.Struct(s)
+	if err == nil {
+		return nil
+	}
+	validationErrors, ok := err.(validator.ValidationErrors)
+	if !ok {
+		return err
+	}
+	fieldErrors := make(FieldError)
+	for _, f := range validationErrors {
+		var field string
+		splitNamespace := strings.Split(f.Namespace(), ".")
+		if len(splitNamespace) > 0 {
+			if len(splitNamespace) > 1 {
+				field = strings.Join(splitNamespace[1:], ".")
+			} else {
+				field = splitNamespace[0]
+			}
+		}
+
+		errMsg := f.Translate(b.translator)
+		fieldErrors[field] = errMsg
+	}
+	return fieldErrors
+}
+
+func (b *Builder) initializeTranslator() ut.Translator {
+	universalTranslator := ut.New(en.New(), en.New())
+	translator, _ := universalTranslator.GetTranslator(defaultLocale)
```

Comment:
Is the second return param an error? Why do we ignore it?

Reply:
I don't meddle much in this code since it is a working code on Dexter so far, but I will try to handle the error and see if something breaks.

Reply:
apparently the second param is not an error but a `found bool` instead. will ignore it for now since I'm still not sure how to handle it.